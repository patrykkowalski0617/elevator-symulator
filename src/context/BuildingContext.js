import React, { createContext, useState } from "react";

export const BuildingContext = createContext();

const BuildingContextProvider = props => {
    const [numberOfFloors, setNumberOfFloors] = useState(8);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
