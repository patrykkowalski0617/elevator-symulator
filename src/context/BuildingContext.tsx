import React, { createContext, useState } from "react";

interface IContextProps {
    numberOfFloors: number;
    setNumberOfFloors: (n: number) => void;
    numberOfCars: number;
    setNnmberOfCars: (n: number) => void;
    automotion: boolean;
    setAutomotion: (b: boolean) => void;
    creatingStickMan: boolean;
    setCreatingStickMan: (arr: boolean) => void;
}

export const BuildingContext = createContext({} as IContextProps);

const BuildingContextProvider = (props: { children: React.ReactNode }) => {
    const [numberOfFloors, setNumberOfFloors] = useState(11);
    const howManyCars = Math.floor(numberOfFloors / 5);
    const [numberOfCars, setNnmberOfCars] = useState(howManyCars);
    const [automotion, setAutomotion] = useState(false);
    const [creatingStickMan, setCreatingStickMan] = useState<boolean>(false);

    return (
        <BuildingContext.Provider
            value={{
                numberOfFloors,
                setNumberOfFloors,
                numberOfCars,
                setNnmberOfCars,
                automotion,
                setAutomotion,
                creatingStickMan,
                setCreatingStickMan
            }}
        >
            {props.children}
        </BuildingContext.Provider>
    );
};

export default BuildingContextProvider;
