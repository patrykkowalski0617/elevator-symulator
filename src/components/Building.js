import React, { useRef, useContext } from "react";
import { BuildingContext } from "../context/BuildingContext";
import Floors from "./Floors";
import styled from "styled-components";
import Shaft from "./Shaft";

const BuildingStyled = styled.div`
    height: 90vh;
    width: 70vw;
    max-width: 600px;
    min-width: 320px;
    margin: 5vh auto;
    background: #000;
    position: relative;
    display: flex;
    flex-flow: row wrap;
    border-style: solid;
    border-width: 6px 4px;
`;

export default function Building() {
    const { numberOfFloors, numberOfCars } = useContext(BuildingContext);

    const buildingRef = useRef();

    // Sizes
    const carWidth = 10;
    const floorWidth = (100 - carWidth * numberOfCars) / 2;

    return (
        <>
            <BuildingStyled ref={buildingRef}>
                <Floors role={"enter-floor"} floorWidth={floorWidth}></Floors>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <Floors role={"exit-floor"} floorWidth={floorWidth}></Floors>
            </BuildingStyled>
        </>
    );
}
