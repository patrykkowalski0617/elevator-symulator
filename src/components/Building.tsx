import React, { useContext } from "react";
import { BuildingContext } from "../context/BuildingContext";
import Floors from "./Floors";
import styled from "styled-components";
import Shaft from "./Shaft";
import StickManForm from "./StickManForm";

const BuildingStyled = styled.div`
    height: 90vh;
    width: 70vw;
    max-width: 600px;
    min-width: 320px;
    margin: 5vh auto;
    background: #000;
    position: relative;
    display: flex;
    flex-flow: row wrap;
    border-style: solid;
    border-width: 6px 4px;
`;

const Building: React.FC = () => {
    const { numberOfFloors, numberOfCars, creatingStickMan } = useContext(
        BuildingContext
    );

    // Sizes
    const carWidth: number = 10;
    const floorWidth: number = (100 - carWidth * numberOfCars) / 2;

    return (
        <>
            {typeof creatingStickMan === "number" ? (
                <StickManForm floorNumber={creatingStickMan} />
            ) : null}
            <BuildingStyled>
                <Floors role={"enter-floor"} floorWidth={floorWidth}></Floors>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <Floors role={"exit-floor"} floorWidth={floorWidth}></Floors>
            </BuildingStyled>
        </>
    );
};

export default Building;
