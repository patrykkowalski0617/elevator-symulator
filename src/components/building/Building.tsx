import React, { useContext } from "react";
import { BuildingContext } from "../../context/BuildingContext";
import FloorsStack from "../floors_stack/FloorsStack";
import { BuildingStyled } from "./BuildingStyled";
import Shaft from "../shaft/Shaft";
import StickManForm from "../form_stickman/FormStickMan";

/**
 * TO DO
 * @Sizes to be removed and used from BuildingContext
 */
const Building: React.FC = () => {
    const { numberOfFloors, numberOfCars, creatingStickMan } = useContext(
        BuildingContext
    );

    // Sizes
    const carWidth: string = "64px";
    const floorWidth: string = ` calc((100% - ${carWidth} * ${numberOfCars}) / 2)`;

    return (
        <>
            {typeof creatingStickMan === "number" ? (
                <StickManForm floorNumber={creatingStickMan} />
            ) : null}
            <BuildingStyled>
                <FloorsStack
                    role={"enter-floor"}
                    floorWidth={floorWidth}
                ></FloorsStack>
                <Shaft
                    numberOfCars={numberOfCars}
                    carWidth={carWidth}
                    numberOfFloors={numberOfFloors}
                ></Shaft>
                <FloorsStack
                    role={"exit-floor"}
                    floorWidth={floorWidth}
                ></FloorsStack>
            </BuildingStyled>
        </>
    );
};

export default Building;
