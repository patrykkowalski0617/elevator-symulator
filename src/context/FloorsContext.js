import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const FloorsContext = createContext();

const FloorsContextProvider = props => {
    const { numberOfFloors } = useContext(BuildingContext);

    const [floorsWaitingForCar, setFloorsWaitingForCar] = useState(
        Array(numberOfFloors).fill(false)
    );

    const addFloorWaitingForCar = floorNumber => {
        const _floorsWaitingForCar = [...floorsWaitingForCar];
        _floorsWaitingForCar.splice(floorNumber, 1, true);
        setFloorsWaitingForCar([..._floorsWaitingForCar]);
    };

    return (
        <FloorsContext.Provider
            value={{
                floorsWaitingForCar,
                addFloorWaitingForCar,
                setFloorsWaitingForCar
            }}
        >
            {props.children}
        </FloorsContext.Provider>
    );
};

export default FloorsContextProvider;
