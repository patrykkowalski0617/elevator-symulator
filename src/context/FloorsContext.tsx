import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

interface IContextProps {
    floorsWaitingForCar: boolean[];
    addFloorWaitingForCar: (floorNumber: number) => void;
    setFloorsWaitingForCar: (arr: boolean[]) => void;
    allStickMansDestinations: number[][];
    addAllStickMansDestinations: (
        floorNumber: number,
        howMany: number,
        destination: number
    ) => void;
}

export const FloorsContext = createContext({} as IContextProps);

const FloorsContextProvider = (props: { children: React.ReactNode }) => {
    const { numberOfFloors } = useContext(BuildingContext);

    const [floorsWaitingForCar, setFloorsWaitingForCar] = useState(
        Array(numberOfFloors).fill(false)
    );

    const [allStickMansDestinations, setallStickMansDestinations] = useState<
        number[][]
    >(Array(numberOfFloors).fill([]));

    const addFloorWaitingForCar = (floorNumber: number) => {
        const _floorsWaitingForCar = [...floorsWaitingForCar];
        _floorsWaitingForCar.splice(floorNumber, 1, true);
        setFloorsWaitingForCar([..._floorsWaitingForCar]);
    };

    const addAllStickMansDestinations = (
        floorNumber: number,
        howMany: number,
        destination: number
    ) => {
        const _allStickMansDestinations: number[][] = [
            ...allStickMansDestinations
        ];
        const newStickMans: number[] = [];
        // push number of destinations for each new StickMan
        for (let i = 0; i < howMany; i++) {
            newStickMans.push(destination);
        }

        const replaceStickMans = newStickMans.concat(
            _allStickMansDestinations[floorNumber]
        );
        _allStickMansDestinations.splice(floorNumber, 1, replaceStickMans);

        setallStickMansDestinations([..._allStickMansDestinations]);
    };

    return (
        <FloorsContext.Provider
            value={{
                floorsWaitingForCar,
                addFloorWaitingForCar,
                setFloorsWaitingForCar,
                allStickMansDestinations,
                addAllStickMansDestinations
            }}
        >
            {props.children}
        </FloorsContext.Provider>
    );
};

export default FloorsContextProvider;
