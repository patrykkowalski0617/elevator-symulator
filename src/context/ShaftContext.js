import React, { createContext, useState, useContext } from "react";
import { BuildingContext } from "./BuildingContext";

export const ShaftContext = createContext();

const ShaftContextProvider = props => {
    const { numberOfCars } = useContext(BuildingContext);
    const initArr = val => Array(numberOfCars).fill(val);

    const [carCurrenFloorCon, setCarCurrenFloorCon] = useState(initArr(0)); // next floor on way or floor that has been just reached, but not floor which has been just left

    // console.log(carCurrenFloorCon);

    return (
        <ShaftContext.Provider
            value={{
                carCurrenFloorCon,
                setCarCurrenFloorCon
            }}
        >
            {props.children}
        </ShaftContext.Provider>
    );
};

export default ShaftContextProvider;
