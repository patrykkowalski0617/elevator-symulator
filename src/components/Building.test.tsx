import React from "react";
import Building from "./Building";
import renderer from "react-test-renderer";
import ShaftContext from "../context/ShaftContext";
import BuildingContext from "../context/BuildingContext";
import FloorsContext from "../context/FloorsContext";

test("Building - render and match snapshot", () => {
    const tree = renderer.create(
        <BuildingContext>
            <ShaftContext>
                <FloorsContext>
                    <Building></Building>
                </FloorsContext>
            </ShaftContext>
        </BuildingContext>
    );
    expect(tree.toJSON()).toMatchSnapshot();
});
