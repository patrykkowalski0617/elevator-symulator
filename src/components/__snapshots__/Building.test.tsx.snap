// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building - render and match snapshot 1`] = `
<div
  className="sc-AxmLO cQCEIO"
>
  <div
    className="sc-AxhCb hJpnPK"
    role="enter-floor"
  >
    <div
      className="sc-AxjAm ghBzSx"
    >
      <button
        data-floor-number={10}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm DgnVW"
    >
      <button
        data-floor-number={9}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm euAsXx"
    >
      <button
        data-floor-number={8}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm ljETuG"
    >
      <button
        data-floor-number={7}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm hIZreI"
    >
      <button
        data-floor-number={6}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm gQikiy"
    >
      <button
        data-floor-number={5}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm eYjwYl"
    >
      <button
        data-floor-number={4}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm hKmDaX"
    >
      <button
        data-floor-number={3}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm kdccyK"
    >
      <button
        data-floor-number={2}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm fwPPRP"
    >
      <button
        data-floor-number={1}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
    <div
      className="sc-AxjAm hOHeui"
    >
      <button
        data-floor-number={0}
        onClick={[Function]}
        style={Object {}}
      >
        Call car
      </button>
    </div>
  </div>
  <div
    className="sc-Axmtr bKgxpp"
  >
    <div
      className="sc-AxhUy gqHuyY"
      style={
        Object {
          "transform": "translateY(calc(-0%))",
        }
      }
    >
      <p
        className="sc-AxgMl bJzgAf"
      >
        C: 
        0
        , T: 
      </p>
      <div
        className="sc-AxheI bBMfsN"
      />
      <div
        className="sc-AxheI iIQMNt"
      />
    </div>
  </div>
  <div
    className="sc-Axmtr bKgxpp"
  >
    <div
      className="sc-AxhUy gqHuyY"
      style={
        Object {
          "transform": "translateY(calc(-0%))",
        }
      }
    >
      <p
        className="sc-AxgMl bJzgAf"
      >
        C: 
        1
        , T: 
      </p>
      <div
        className="sc-AxheI bBMfsN"
      />
      <div
        className="sc-AxheI iIQMNt"
      />
    </div>
  </div>
  <div
    className="sc-AxhCb hJpnPK"
    role="exit-floor"
  >
    <div
      className="sc-AxjAm ghBzSx"
    >
      <p>
        10
      </p>
    </div>
    <div
      className="sc-AxjAm DgnVW"
    >
      <p>
        9
      </p>
    </div>
    <div
      className="sc-AxjAm euAsXx"
    >
      <p>
        8
      </p>
    </div>
    <div
      className="sc-AxjAm ljETuG"
    >
      <p>
        7
      </p>
    </div>
    <div
      className="sc-AxjAm hIZreI"
    >
      <p>
        6
      </p>
    </div>
    <div
      className="sc-AxjAm gQikiy"
    >
      <p>
        5
      </p>
    </div>
    <div
      className="sc-AxjAm eYjwYl"
    >
      <p>
        4
      </p>
    </div>
    <div
      className="sc-AxjAm hKmDaX"
    >
      <p>
        3
      </p>
    </div>
    <div
      className="sc-AxjAm kdccyK"
    >
      <p>
        2
      </p>
    </div>
    <div
      className="sc-AxjAm fwPPRP"
    >
      <p>
        1
      </p>
    </div>
    <div
      className="sc-AxjAm hOHeui"
    >
      <p>
        0
      </p>
    </div>
  </div>
</div>
`;
